File: zlh/LeetCode/2023-11-19/101. 对称二叉树.ts
function isSymmetric(root: TreeNode | null): boolean {
  if (!root) return true;
  return compareNode(root.left, root.right);
}

const compareNode = (
  left: TreeNode | null,
  right: TreeNode | null
): boolean => {
  if (left === null && right !== null) return false;
  else if (left !== null && right === null) return false;
  else if (left === null && right === null) return true;
  else if (left.val !== right.val) return false;

  let outside: boolean = compareNode(left.left, right.right);
  let inside: boolean = compareNode(left.right, right.left);

  return outside && inside;
};

File: zjj/101. 对称二叉树.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode[TreeNode]) -> bool:
        if not root:
            return True

        def dfs(L, R):
            if not L and not R:
                return True
            if not L or not R or L.val != R.val:
                return False
            return dfs(L.left, R.right) and dfs(L.right, R.left)

        return dfs(root.left, root.right)
